pysrc/bytewax/connectors/kafka/operators.py

```diff
@@ -0,0 +1,289 @@
+
+    Returns an object with two attributes:
+        - .oks: A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+        - .errors: A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    """
+
+    # Make sure the first KafkaMessage in the return type's Union represents
+    # the `oks` stream and the second one the `errs` stream for the
+    # `_kafka_error_split` operator.
+    def shim_mapper(
+        msg: KafkaMessage[K, V]
+    ) -> Union[KafkaMessage[K2, V], KafkaMessage[K, V]]:
+        try:
+            key = deserializer.de(msg.key)
+            return msg.with_key(key)
+        except Exception as e:
+            return msg.with_error(KafkaError(KafkaError._KEY_DESERIALIZATION, f"{e}"))
+
+    return op.map("map", up, shim_mapper).then(_kafka_error_split, "split")
+
+
+@operator
+def deserialize_value(
+    step_id: str,
+    up: Stream[KafkaMessage[K, V]],
+    deserializer: SchemaDeserializer[V, V2],
+) -> KafkaStreams[K, V2, K, V]:
+    """Deserialize the value of a KafkaMessage using the provided SchemaDeserializer.
+
+    Returns an object with two attributes:
+        - .oks: A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+        - .errors: A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    """
+
+    # Make sure the first KafkaMessage in the return type's Union represents
+    # the `oks` stream and the second one the `errs` stream for the
+    # `_kafka_error_split` operator.
+    def shim_mapper(
+        msg: KafkaMessage[K, V]
+    ) -> Union[KafkaMessage[K, V2], KafkaMessage[K, V]]:
+        try:
+            value = deserializer.de(msg.value)
+            return msg.with_value(value)
+        except Exception as e:
+            return msg.with_error(KafkaError(KafkaError._VALUE_DESERIALIZATION, f"{e}"))
+
+    return op.map("map", up, shim_mapper).then(_kafka_error_split, "split_err")
+
+
+@operator
+def deserialize(
+    step_id: str,
+    up: Stream[KafkaMessage[K, V]],
+    *,
+    key_deserializer: SchemaDeserializer[K, K2],
+    val_deserializer: SchemaDeserializer[V, V2],
+) -> KafkaStreams[K2, V2, Union[K, K2], Union[V, V2]]:
+    """Serialize both keys and values with the given serializers.
+
+    Returns an object with two attributes:
+        - .oks: A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+        - .errors: A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    A message will be put in .errors even if only one of the deserializers fail.
+    """
+    keys = deserialize_key("de-key", up, key_deserializer)
+    values = deserialize_value("de-val", keys.oks, val_deserializer)
+    errors = cast(
+        # Either of key and value deserialization might have failed in the
+        # errors stream. Since we don't know better, just Union the possible
+        # output types
+        Stream[KafkaMessage[Union[K, K2], Union[V, V2]]],
+        op.merge("merg-errors", keys.errs, values.errs),
+    )
```

Comment:
Right, I can use a separate mapper in this operator rather than chaining the existing one, so that if any of the 2 fail we keep the original message. Fixed